src/data_types/w3c/credential.rs

```diff
@@ -0,0 +1,367 @@
+                    match raw_value {
+                        CredentialAttributeValue::Attribute(raw_value) => {
+                            cred_values.add_raw(attribute, raw_value)?
+                        }
+                        value => {
+                            return Err(err_msg!(
+                                "Encoding is not supported for credential value {:?}",
+                                value
+                            ));
+                        }
+                    }
+                }
+                Ok(cred_values.into())
+            }
+            encoding => Err(err_msg!(
+                "Credential values encoding {:?} is not supported",
+                encoding
+            )),
+        }
+    }
+}
+
+#[derive(Debug, Clone, Default, PartialEq, Eq, Serialize, Deserialize)]
+pub struct CredentialStatus {
+    #[serde(rename = "type")]
+    pub type_: CredentialStatusType,
+    pub id: RevocationRegistryDefinitionId,
+}
+
+#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
+pub enum CredentialStatusType {
+    #[serde(rename = "AnonCredsCredentialStatusList2023")]
+    AnonCredsCredentialStatusList2023,
+    Other(String),
+}
+
+impl Default for CredentialStatusType {
+    fn default() -> Self {
+        CredentialStatusType::AnonCredsCredentialStatusList2023
+    }
+}
+
+impl CredentialStatus {
+    pub fn new(id: RevocationRegistryDefinitionId) -> CredentialStatus {
+        CredentialStatus {
+            type_: CredentialStatusType::AnonCredsCredentialStatusList2023,
+            id,
+        }
+    }
+}
+
+#[derive(Debug, Clone, Default, PartialEq, Eq, Serialize, Deserialize)]
+#[serde(rename_all = "camelCase")]
+pub struct CredentialSchema {
+    #[serde(rename = "type")]
+    pub type_: CredentialSchemaType,
+    pub definition: CredentialDefinitionId,
+    pub schema: SchemaId,
+    #[serde(default)]
+    pub encoding: CredentialValuesEncoding,
+}
+
+impl CredentialSchema {
+    pub fn new(
+        schema: SchemaId,
+        definition: CredentialDefinitionId,
+        encoding: CredentialValuesEncoding,
+    ) -> CredentialSchema {
+        CredentialSchema {
+            type_: CredentialSchemaType::AnonCredsDefinition,
+            definition,
+            schema,
+            encoding,
+        }
+    }
+}
+
+#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
+pub enum CredentialSchemaType {
+    #[serde(rename = "AnonCredsDefinition")]
+    AnonCredsDefinition,
+}
+
+impl Default for CredentialSchemaType {
+    fn default() -> Self {
+        CredentialSchemaType::AnonCredsDefinition
+    }
+}
+
+impl W3CCredential {
+    pub fn new() -> W3CCredential {
+        W3CCredential {
+            context: ANONCREDS_CONTEXTS.clone(),
+            type_: ANONCREDS_CREDENTIAL_TYPES.clone(),
+            issuance_date: Utc::now(),
+            proof: OneOrMany::Many(Vec::new()),
+            ..Default::default()
+        }
+    }
+
+    pub fn set_id(&mut self, id: URI) {
```

Comment:
Not the biggest fan of this API right now, but I am also unsure what can be done to improve it.